domain/secret/service/watcher.go

```diff
@@ -155,6 +152,269 @@ func (s *WatchableService) WatchSecretRevisionsExpiryChanges(_ context.Context,
 	return newSecretStringWatcher(w, s.logger, processChanges)
 }
 
+type obsoleteWatcher struct {
+	catacomb       catacomb.Catacomb
+	logger         logger.Logger
+	state          State
+	watcherFactory WatcherFactory
+
+	appOwners  domainsecret.ApplicationOwners
+	unitOwners domainsecret.UnitOwners
+
+	// secretWatcher watches for secret creation and deletion changes.
+	// It returns the secret URI.
+	secretWatcher watcher.StringsWatcher
+	// obsoleteRevisionWatcher watches for obsolete revision update changes.
+	// It returns the obsolete revision UUID.
+	revisionWatcher watcher.StringsWatcher
+
+	// knownSecretURIs is a set of currently owned secret URIs.
+	// Tracking this set allows us to identify if a deletion event corresponds to a previously owned secret.
+	// When a deletion event is received, the secret data is no longer available in the database,
+	// so we cannot query the database to determine if the secret was previously owned.
+	knownSecretURIs set.Strings
+
+	out chan []string
+}
+
+func newObsoleteWatcher(
+	state State,
+	watcherFactory WatcherFactory,
+	logger logger.Logger,
+	owners ...CharmSecretOwner,
+) (*obsoleteWatcher, error) {
+	appOwners, unitOwners := splitCharmSecretOwners(owners...)
+	if len(appOwners) == 0 && len(unitOwners) == 0 {
+		return nil, internalerrors.Errorf("at least one owner must be provided for the obsolete watcher")
+	}
+
+	w := &obsoleteWatcher{
+		logger:          logger,
+		state:           state,
+		watcherFactory:  watcherFactory,
+		appOwners:       appOwners,
+		unitOwners:      unitOwners,
+		knownSecretURIs: set.NewStrings(),
+		out:             make(chan []string),
+	}
+
+	if err := w.init(); err != nil {
+		return nil, internalerrors.Capture(err)
+	}
+
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{w.secretWatcher, w.revisionWatcher},
+	})
+	return w, internalerrors.Capture(err)
+}
+
+func (w *obsoleteWatcher) init() error {
+	var err error
+	table, query := w.state.InitialWatchStatementForSecretMatadata(w.appOwners, w.unitOwners)
```

Comment:
Could this be called `InitialWatchStatementForOwnedSecrets`

I think that makes more sense  it does to me anyway.